File: Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/FileCacheDownloader.java
/*
 * Kuroba - *chan browser https://github.com/Adamantcheese/Kuroba/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.adamantcheese.chan.core.cache;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.AnyThread;
import androidx.annotation.MainThread;
import androidx.annotation.WorkerThread;

import com.github.adamantcheese.chan.Chan;
import com.github.adamantcheese.chan.core.di.NetModule;
import com.github.adamantcheese.chan.core.settings.ChanSettings;
import com.github.adamantcheese.chan.utils.Logger;

import java.io.Closeable;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

import okhttp3.Call;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okhttp3.internal.Util;
import okio.Buffer;
import okio.BufferedSink;
import okio.Okio;
import okio.Source;

public class FileCacheDownloader implements Runnable {
    private static final String TAG = "FileCacheDownloader";
    private static final long BUFFER_SIZE = 32768;
    private static final long NOTIFY_SIZE = BUFFER_SIZE * 8;

    private final String url;
    private final File output;
    private final Handler handler;

    // Main thread only.
    private final Callback callback;
    private final List<FileCacheListener> listeners = new ArrayList<>();
    private final OkHttpClient okHttpClient;

    // Main and worker thread.
    private AtomicBoolean running = new AtomicBoolean(false);
    private AtomicBoolean cancel = new AtomicBoolean(false);

    // Worker thread.
    private Call call;
    private ResponseBody body;

    public FileCacheDownloader(Callback callback, String url, File output) {
        this.callback = callback;
        this.url = url;
        this.output = output;

        okHttpClient = new OkHttpClient().newBuilder()
                .proxy(ChanSettings.getProxy())
                .build();

        handler = new Handler(Looper.getMainLooper());
    }

    @MainThread
    public String getUrl() {
        return url;
    }

    @MainThread
    public void addListener(FileCacheListener callback) {
        listeners.add(callback);
    }

    /**
     * Cancel this download.
     */
    @MainThread
    public void cancel() {
        if (cancel.compareAndSet(false, true)) {
            // Did not start running yet, mark finished here.
            if (!running.get() && callback != null) {
                callback.downloaderFinished(this);
            }
        }
    }

    @AnyThread
    private void log(String message) {
        Logger.d(TAG, logPrefix() + message);
    }

    @AnyThread
    private void log(String message, Exception e) {
        Logger.e(TAG, logPrefix() + message, e);
    }

    private String logPrefix() {
        return "[" + url.substring(0, Math.min(url.length(), 45)) + "] ";
    }

    @Override
    @WorkerThread
    public void run() {
        log("start");
        running.set(true);
        execute();
    }

    @WorkerThread
    private void execute() {
        Closeable sourceCloseable = null;
        Closeable sinkCloseable = null;

        try {
            checkCancel();

            ResponseBody body = getBody();

            Source source = body.source();
            sourceCloseable = source;

            BufferedSink sink = Okio.buffer(Okio.sink(output));
            sinkCloseable = sink;

            checkCancel();

            log("got input stream");

            pipeBody(source, sink);

            log("done");

            long fileLen = output.length();

            handler.post(() -> {
                if (callback != null) {
                    callback.downloaderAddedFile(fileLen);
                    callback.downloaderFinished(this);
                }
                for (FileCacheListener callback : listeners) {
                    callback.onSuccess(output);
                    callback.onEnd();
                }
            });
        } catch (IOException e) {
            boolean isNotFound = false;
            boolean cancelled = false;
            if (e instanceof HttpCodeIOException) {
                int code = ((HttpCodeIOException) e).code;
                log("exception: http error, code: " + code, e);
                isNotFound = code == 404;
            } else if (e instanceof CancelException) {
                // Don't log the stack.
                log("exception: cancelled");
                cancelled = true;
            } else {
                log("exception", e);
            }

            final boolean finalIsNotFound = isNotFound;
            final boolean finalCancelled = cancelled;
            handler.post(() -> {
                purgeOutput();
                for (FileCacheListener callback : listeners) {
                    if (finalCancelled) {
                        callback.onCancel();
                    } else {
                        callback.onFail(finalIsNotFound);
                    }

                    callback.onEnd();
                }
                if (callback != null) {
                    callback.downloaderFinished(this);
                }
            });
        } finally {
            Util.closeQuietly(sourceCloseable);
            Util.closeQuietly(sinkCloseable);

            if (call != null) {
                call.cancel();
            }

            if (body != null) {
                Util.closeQuietly(body);
            }
        }
    }

    @WorkerThread
    private ResponseBody getBody() throws IOException {
        Request request = new Request.Builder()
                .url(url)
                .header("User-Agent", NetModule.USER_AGENT)
                .build();

        call = okHttpClient.newCall(request);
        Response response = call.execute();
        
        if (!response.isSuccessful()) {
            throw new HttpCodeIOException(response.code());
        }

        checkCancel();

        body = response.body();
        if (body == null) {
            throw new IOException("body == null");
        }

        checkCancel();

        return body;
    }

    @WorkerThread
    private void pipeBody(Source source, BufferedSink sink) throws IOException {
        long contentLength = body.contentLength();

        long read;
        long total = 0;
        long notifyTotal = 0;

        Buffer buffer = new Buffer();

        while ((read = source.read(buffer, BUFFER_SIZE)) != -1) {
            sink.write(buffer, read);
            total += read;

            if (total >= notifyTotal + NOTIFY_SIZE) {
                notifyTotal = total;
                log("progress " + (total / (float) contentLength));
                postProgress(total, contentLength <= 0 ? total : contentLength);
            }

            checkCancel();
        }

        Util.closeQuietly(source);
        Util.closeQuietly(sink);

        call = null;
        Util.closeQuietly(body);
        body = null;
    }

    @WorkerThread
    private void checkCancel() throws IOException {
        if (cancel.get()) {
            throw new CancelException();
        }
    }

    @WorkerThread
    private void purgeOutput() {
        if (output.exists()) {
            final boolean deleteResult = output.delete();

            if (!deleteResult) {
                log("could not delete the file in purgeOutput");
            }
        }
    }

    @WorkerThread
    private void postProgress(final long downloaded, final long total) {
        handler.post(() -> {
            for (FileCacheListener callback : listeners) {
                callback.onProgress(downloaded, total);
            }
        });
    }

    private static class CancelException extends IOException {
    }

    private static class HttpCodeIOException extends IOException {
        private int code;

        public HttpCodeIOException(int code) {
            this.code = code;
        }
    }

    public interface Callback {
        void downloaderFinished(FileCacheDownloader fileCacheDownloader);

        void downloaderAddedFile(long fileLen);
    }
}

File: Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/FileCacheDownloader.kt
/*
 * Kuroba - *chan browser https://github.com/Adamantcheese/Kuroba/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.adamantcheese.chan.core.cache

import android.os.Handler
import android.os.Looper
import androidx.annotation.AnyThread
import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import com.github.adamantcheese.chan.Chan
import com.github.adamantcheese.chan.core.di.NetModule
import com.github.adamantcheese.chan.utils.BackgroundUtils
import com.github.adamantcheese.chan.utils.Logger
import com.github.k1rakishou.fsaf.FileManager
import com.github.k1rakishou.fsaf.file.RawFile
import okhttp3.Call
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.ResponseBody
import okhttp3.internal.closeQuietly
import okio.*
import java.io.Closeable
import java.io.IOException
import java.io.OutputStream
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

class FileCacheDownloader(
        private val fileManager: FileManager,
        // Main thread only.
        private val callback: Callback?,
        private val output: RawFile,
        val url: String
) : Runnable {
    private val handler: Handler = Handler(Looper.getMainLooper())
    private val listeners = ArrayList<FileCacheListener>()

    // Main and worker thread.
    private val running = AtomicBoolean(false)
    private val cancel = AtomicBoolean(false)

    // When encountering an IOException during the attempt to download a file instead of immediately
    // terminating the downloading we will try to download it couple of more times (3 times max).
    // If we are still unable to download the file after all of the attempts (no internet connection)
    // we give up and throw the latest exception. This should fix the annoying "Saved image failed"
    // error when downloading thread albums.
    //
    // The reason is actually SocketTimeoutException is the server forcefully drops
    // the connection if it detects that we are downloading files?
    private val retryAttempts = AtomicInteger(3)
    private val retryTimeoutsSeconds = arrayOf(5L, 2L, 1L)

    @MainThread
    fun addListener(callback: FileCacheListener) {
        BackgroundUtils.ensureMainThread()

        listeners.add(callback)
    }

    /**
     * Cancel this download.
     */
    @MainThread
    fun cancel() {
        BackgroundUtils.ensureMainThread()

        if (cancel.compareAndSet(false, true)) {
            // Did not start running yet, mark finished here.
            if (!running.get() && callback != null) {
                callback.downloaderFinished(this)
            }
        }
    }

    @AnyThread
    private fun log(message: String) {
        Logger.d(TAG, logPrefix() + message)
    }

    @AnyThread
    private fun log(message: String, e: Exception) {
        Logger.e(TAG, logPrefix() + message, e)
    }

    private fun logPrefix(): String {
        return "[" + url.substring(0, url.length.coerceAtMost(45)) + "] "
    }

    @WorkerThread
    override fun run() {
        BackgroundUtils.ensureBackgroundThread()

        log("start")
        running.set(true)
        execute()
    }

    @WorkerThread
    private fun execute() {
        var sourceCloseable: Closeable? = null
        var sinkCloseable: Closeable? = null
        var outputFileOutputStream: OutputStream? = null
        var call: Call? = null
        var body: ResponseBody? = null

        try {
            BackgroundUtils.ensureBackgroundThread()
            checkCancel()

            val timeout = retryTimeoutsSeconds.getOrElse(retryAttempts.get()) { 0 }
            if (timeout > 0) {
                try {
                    log("Sleeping for $timeout seconds")
                    Thread.sleep(timeout * 1000L)
                } catch (error: InterruptedException) {
                    Thread.currentThread().interrupt()
                }
            }

            checkCancel()

            val startTime = System.currentTimeMillis()
            val (newCall, newBody) = getBody()
            if (newBody == null) {
                throw IOException("body == null")
            }

            call = newCall
            body = newBody

            val source = body.source().also { source ->
                sourceCloseable = source
            }

            if (!fileManager.exists(output) && fileManager.create(output) == null) {
                throw IOException("Couldn't create output file, output = " + output.getFullPath())
            }

            outputFileOutputStream = fileManager.getOutputStream(output)
            if (outputFileOutputStream == null) {
                throw IOException("Couldn't get output file's OutputStream")
            }

            val sink = outputFileOutputStream.sink().buffer().also { sink ->
                sinkCloseable = sink
            }

            checkCancel()

            log("got input stream")
            pipeBody(body, source, sink)

            val deltaTime = System.currentTimeMillis() - startTime
            val fileLength = fileManager.getLength(output)
            val fileSizeInKB = fileLength / 1024

            log("File with size $fileSizeInKB KB was downloaded done in $deltaTime ms")

            postSuccessResult(
                    callback,
                    fileLength
            )

        } catch (e: IOException) {
            BackgroundUtils.ensureBackgroundThread()

            var isNotFound = false
            var canceled = false

            when (e) {
                is HttpCodeIOException -> {
                    val code = e.code
                    log("exception: http error, code: $code", e)
                    isNotFound = code == 404
                }
                is CancelException -> {
                    // Don't log the stack.
                    log("exception: canceled")
                    canceled = true
                }
                else -> {
                    if (retryRequest()) {
                        log("An IOException (" + e.javaClass.simpleName +
                                ") has occurred, there are " + retryAttempts.get() +
                                " retry attempts, running the request again")

                        execute()
                    } else {
                        log("No more retry attempts left, throwing the exception", e)
                        postNetworkError(e)
                    }
                }
            }

            val finalIsNotFound = isNotFound
            val finalCanceled = canceled

            postErrorResult(finalCanceled, finalIsNotFound)
        } finally {
            BackgroundUtils.ensureBackgroundThread()

            sourceCloseable?.closeQuietly()
            sinkCloseable?.closeQuietly()
            outputFileOutputStream?.closeQuietly()
            call?.cancel()
            body?.closeQuietly()
        }
    }

    private fun retryRequest(): Boolean {
        BackgroundUtils.ensureBackgroundThread()
        return retryAttempts.getAndDecrement() > 0
    }

    private fun postNetworkError(error: IOException) {
        BackgroundUtils.ensureBackgroundThread()

        handler.post {
            for (callback in listeners) {
                callback.onNetworkError(error)
            }
        }
    }

    private fun postErrorResult(finalCanceled: Boolean, finalIsNotFound: Boolean) {
        BackgroundUtils.ensureBackgroundThread()
        purgeOutput()

        handler.post {
            for (callback in listeners) {
                if (finalCanceled) {
                    callback.onCancel()
                } else {
                    callback.onFail(finalIsNotFound)
                }

                callback.onEnd()
            }

            callback?.downloaderFinished(this)
        }
    }

    private fun postSuccessResult(callback: Callback?, fileLen: Long) {
        BackgroundUtils.ensureBackgroundThread()

        handler.post {
            if (callback != null) {
                callback.downloaderAddedFile(fileLen)
                callback.downloaderFinished(this)
            }

            listeners.forEach { cb ->
                cb.onSuccess(output)
                cb.onEnd()
            }
        }
    }

    @WorkerThread
    @Throws(IOException::class)
    private fun getBody(): Pair<Call, ResponseBody?> {
        BackgroundUtils.ensureBackgroundThread()

        val request = Request.Builder()
                .url(url)
                .header("User-Agent", NetModule.USER_AGENT)
                .build()

        //we want to use the proxy instance here
        val call = (Chan.injector().instance(OkHttpClient::class.java) as NetModule.ProxiedOkHttpClient)
                .proxiedClient.newCall(request)

        val response = call.execute()
        if (!response.isSuccessful) {
            throw HttpCodeIOException(response.code)
        }

        checkCancel()
        return Pair(call, response.body)
    }

    @WorkerThread
    @Throws(IOException::class)
    private fun pipeBody(body: ResponseBody?, source: Source, sink: BufferedSink) {
        BackgroundUtils.ensureBackgroundThread()

        if (body == null) {
            throw IOException("pipeBody() body == null")
        }

        val contentLength = body.contentLength()

        var read: Long
        var total: Long = 0
        var notifyTotal: Long = 0
        val buffer = Buffer()

        while (true) {
            read = source.read(buffer, BUFFER_SIZE)
            if (read == -1L) {
                break
            }

            sink.write(buffer, read)
            total += read

            if (total >= notifyTotal + NOTIFY_SIZE) {
                notifyTotal = total
                log("progress " + total / contentLength.toFloat())
                postProgress(total, if (contentLength <= 0) total else contentLength)
            }

            checkCancel()
        }

        source.closeQuietly()
        sink.closeQuietly()
        body.closeQuietly()
    }

    @WorkerThread
    @Throws(IOException::class)
    private fun checkCancel() {
        BackgroundUtils.ensureBackgroundThread()

        if (cancel.get()) {
            throw CancelException()
        }
    }

    @WorkerThread
    private fun purgeOutput() {
        BackgroundUtils.ensureBackgroundThread()

        if (fileManager.exists(output)) {
            val deleteResult = fileManager.delete(output)

            if (!deleteResult) {
                log("could not delete the file in purgeOutput")
            }
        }
    }

    @WorkerThread
    private fun postProgress(downloaded: Long, total: Long) {
        BackgroundUtils.ensureBackgroundThread()

        handler.post {
            for (callback in listeners) {
                callback.onProgress(downloaded, total)
            }
        }
    }

    private class CancelException : IOException()
    private class HttpCodeIOException(val code: Int) : IOException()

    interface Callback {
        fun downloaderFinished(fileCacheDownloader: FileCacheDownloader)
        fun downloaderAddedFile(fileLen: Long)
    }

    companion object {
        private const val TAG = "FileCacheDownloader"
        private const val BUFFER_SIZE: Long = 8192
        private const val NOTIFY_SIZE = BUFFER_SIZE * 8
    }
}
